be/src/vec/exec/join/vhash_join_node.cpp

```diff
@@ -177,37 +180,66 @@ struct ProcessHashTableProbe {
 
         std::vector<uint32_t> items_counts(_probe_rows);
         auto& mcol = mutable_block.mutable_columns();
-
-        int right_col_idx = _join_node->_is_right_semi_anti ? 0 : _left_table_data_types.size();
-        int right_col_len = _right_table_data_types.size();
         int current_offset = 0;
+        std::vector<uint32_t> _build_index;
+        _build_index.reserve(1.2 * _batch_size);
 
         for (; _probe_index < _probe_rows;) {
-            // ignore null rows
             if constexpr (ignore_null) {
                 if ((*null_map)[_probe_index]) {
                     items_counts[_probe_index++] = 0;
                     continue;
                 }
             }
-
             int repeat_count = 0;
-            auto find_result =
-                    (*null_map)[_probe_index]
+            if constexpr (is_inner_join) {
+                if (!(*null_map)[_probe_index]) {
+                    auto find_result = key_getter.find_key(hash_table_ctx.hash_table, _probe_index, _arena);
+
+                    if (find_result.is_found()) {
+                        auto& mapped = find_result.get_mapped();
+
+                        if (mapped.get_row_count() == 1) {
+                            ++repeat_count;
+                            _build_index.emplace_back(mapped.row_num);
+                        } else {
+                            if (_probe_index + 2 < _probe_rows)
+                                key_getter.prefetch(hash_table_ctx.hash_table, _probe_index + 2, _arena);
+                            for (auto it = mapped.begin(); it.ok(); ++it) {
+                                ++repeat_count;
+                                _build_index.emplace_back(it->row_num);
```

Comment:
if row_num stored in mapped as continuous array, we can just remember the row_num array’s address here. 
it's no need to copy one by one.